### DISCUSSION TITLE
Rough Plan

### DISCUSSION THREAD

**Discussion by kunitoki** [kunitoki/yup#4] [General]
Platform
- [x] Add support for running on Android with NativeActivity https://github.com/kunitoki/yup/pull/34
- [x] Add support for running on iOS https://github.com/kunitoki/yup/pull/34
- [x] Improved SDL support, more display management
- [ ] Transition to SDL3, and especially to the new [AppCallbacks](https://wiki.libsdl.org/SDL3/SDL_EnterAppMainCallbacks) - blocked by https://github.com/libsdl-org/SDL/issues/12745
- [ ] Improved Emscripten support

Graphics
- [ ] Experiment with rendering in a separate thread
- [x] Improve caching of paths and gradients when possible
- [x] Add support for headless graphics context / no op rive factory
- [ ] Add support for Vulkan surfaces
- [ ] Add support for WebGPU (using Dawn on desktop)
- [x] Some form of SVG support to test the graphics class https://github.com/kunitoki/yup/pull/64
- [ ] Support GLES3.1 on linux (Raspberry PI) https://github.com/kunitoki/yup/issues/49
- [x] Improved cpu and gpu image support
- [ ] Enable post processing shader support

Audio
- [x] Add support for Standalone plugins
- [x] Add support for VST3 plugins https://github.com/kunitoki/yup/pull/44
- [x] Add support for AU plugins
- [x] Audio Plugin Host AUv2, VST3, CLAP (mac)
- [ ] Audio Plugin Host AUv2, VST3, CLAP (windows)
- [ ] Audio Plugin Host AUv2, VST3, CLAP (linux)
- [ ] Re-evaluate clap-wrapper usage
- [ ] Lock free fifo improvement for moving data to/from audio thread
- [x] Improved audio parameters
- [x] Emscripten AudioWorklet support https://github.com/kunitoki/yup/pull/25
- [x] Barebone yup_dsp module https://github.com/kunitoki/yup/pull/71
- [x] FFT and windowing https://github.com/kunitoki/yup/pull/71
- [x] Basic IIR Filters https://github.com/kunitoki/yup/pull/71
- [x] Linkwitz Riley filters for crossover operations https://github.com/kunitoki/yup/pull/71
- [ ] Advanced IIR Filters (Bessel, Legendre, Elliptic, Chebyshev, ...)
- [x] Upsampling and Downsampling Filters https://github.com/kunitoki/yup/pull/97
- [x] FIR Filter
- [x] Convolution https://github.com/kunitoki/yup/pull/75
- [ ] Oscillators/tone generators
- [x] Noise generators https://github.com/kunitoki/yup/pull/71

UI
- [x] Customizable skinning support (ala LookAndFeel, but better) https://github.com/kunitoki/yup/discussions/10 https://github.com/kunitoki/yup/pull/13
- [x] Better slider component (knob, linear, range, ...) https://github.com/kunitoki/yup/pull/70
- [x] Better button components (toggle, text, shape, image, ...)
- [x] Midi keyboard component https://github.com/kunitoki/yup/pull/70
- [ ] Oscilloscope
- [x] Filters visualisation https://github.com/kunitoki/yup/pull/71
- [x] Spectrum analyser https://github.com/kunitoki/yup/pull/71
- [x] Improved font support, better performance
- [x] Better font layouting
- [x] Variable fonts axis manipulation
- [x] Embedded fallback font
- [x] TextEditor and Label components https://github.com/kunitoki/yup/pull/16 https://github.com/kunitoki/yup/pull/55
- [x] Image drawing support
- [x] Popup menus https://github.com/kunitoki/yup/pull/57
- [x] File choosers https://github.com/kunitoki/yup/pull/61

Data Models
- [x] Dedicated module to data modelling https://github.com/kunitoki/yup/pull/15
- [x] Undo manager support https://github.com/kunitoki/yup/pull/15
- [x] DataTree support https://github.com/kunitoki/yup/pull/74
- [x] DataTree query support https://github.com/kunitoki/yup/pull/74
- [x] DataTree schema validation https://github.com/kunitoki/yup/pull/74
- [x] CachedValue class to hold typesafe references to DataTree properties https://github.com/kunitoki/yup/pull/74
- [x] Clipboard support https://github.com/kunitoki/yup/pull/55

Artboard
- [x] Shared rive file across multiple Artboards components https://github.com/kunitoki/yup/pull/43
- [x] Write inputs into state machines https://github.com/kunitoki/yup/pull/43
- [x] Improved artboard placement
- [ ] Read inputs from state machines
- [x] React on events from state machines https://github.com/kunitoki/yup/pull/43
- [ ] Data binding / View model support
- [x] Multi artboard Component support
- [ ] Static artboard Component for smaller and faster render of static vectors (svg alternative)
- [ ] Query artboard dimensions and named nodes

AI
- [ ] Local LLM requests with tooling https://github.com/kunitoki/yup/pull/99
- [ ] MCP Client and Servers https://github.com/kunitoki/yup/pull/99
- [ ] Python support https://github.com/kunitoki/yup/pull/99
- [ ] Embeddings
- [ ] ONNX wrapper

Misc
- [x] More graphics demos
- [ ] Fix includes in each header to help IDEs and intellisense (but also AI tools)
- [ ] Proper demo app
- [ ] External standalone plugin to show the capabilities
- [ ] Templates repositories for standalone and plugin development

Utilities
- [x] Profiling with perfetto sdk https://github.com/kunitoki/yup/pull/20
- [x] Codecov coverage integration https://github.com/kunitoki/yup/pull/54
- [x] Integrate `yup_python` module from [popsicle](https://github.com/kunitoki/popsicle) https://github.com/kunitoki/yup/pull/65
- [x] Integrate `pluginval` in the build system https://github.com/kunitoki/yup/pull/67

**Comment by adamski** [comment]
@kunitoki what do you think would be a sensible order to address these tasks?

**Comment by kunitoki** [comment]
I will focus on getting a standalone plugin first, so we can easily settle a barebone api for creating audio apps with parameters, presets and stuff. Once that is settled, more basic widgets needs to be provided, and i'm already working on rive widgets too (sliders, buttons, midi keyboard...) and more examples should be done, eventually showcasing an audio plugin too

**Comment by kunitoki** [comment]
I'm currently working on the Component themes support before i will start working on more components. Having that one setup nicely and future proof will allow us to move faster and avoid to change the design multiple times as we go. 

See https://github.com/kunitoki/yup/discussions/10 for more information and ideas. 

*You can definately help shaping it !*

**Comment by kunitoki** [comment]
Closing this as i will be working on individual roadmaps
